### ISSUE TITLE
Issue with Instance Weights in XGBoost for Learning to Rank

### ISSUE THREAD

**Issue by HemakshiShardha** [dmlc/xgboost#11633]
I’m currently using XGBoost version 1.3.1 to train a Learning to Rank (LTR) model, and I’m having an issue with instance weights. Here’s my setup:

- I have training and validation data in the same directory, along with two separate files: one for instance weights (.weight file) and another for group information (.group file).
- The weights file contains event-level weights, with one weight assigned to each event.
- The group file contains number of instances in each event, with one value assigned to each event.
- I’ve confirmed that weights are being passed to the DMatrix by printing them using the .get_weight() method.

The issue is that when I train two models—one with weights and one without—the validation NDCG scores for both models are almost identical, and I’m wondering if the model is even using the weights. To further investigate, I’ve even tried using extreme values for the weights, but the NDCG scores are still similar for the model with and without extreme weights.

Is this expected behavior, or is there something I might be missing? Has anyone else encountered this issue, or could there be a bug with instance weights handling in this version?

Thanks

**Comment by trivialfis** [comment]
Hi, could you please test the latest XGBoost, and share a reproducible example if the issue persists?

**Comment by HemakshiShardha** [comment]
Hi,  I am using XGBoost 3.0.4. I have a quick question.
Are ndcg@5 and ndcg@10 weight-agnostic or should I create a weight-aware metric to analyze the impact of providing instance weights to my LTR model? I am trying to understand if I should expect to see quite different nDCG values if I train a model with no weights and another model with extreme weights? Thanks

**Comment by trivialfis** [comment]
Weight has effect for all objectives, including NDCG The special case for learning to rank is that weight is per-group instead of per-sample.

**Comment by HemakshiShardha** [comment]
Here is an example where the NDCG scores remain nearly identical even with extreme weights. 80% of the events in training data are from one department and 20% are from another. I want to give weights such that the model learns more about the department which has only 20% events in the training data

```
import numpy as np
import xgboost as xgb # using version 3.0.4
import tempfile
import os
from sklearn.metrics import ndcg_score

np.random.seed(424)

print("XGBoost version:", xgb.__version__)

# Generate training data
n_queries_train = 10000 # number of events in training table
n_features = 10 # num of features

X_train_list = []
y_train_list = []
groups_train = []
group_labels_train = []

for i in range(n_queries_train):
    group = 0 if i < 80 else 1 # 80% of the events are from one department and rest 20% are from another
    group_labels_train.append(group)
    
    n_docs = np.random.randint(5, 15)
    groups_train.append(n_docs)
    
    X_query = np.random.randn(n_docs, n_features)
    y_query = np.random.randint(0, 5, n_docs)
    
    X_train_list.append(X_query)
    y_train_list.append(y_query)

X_train = np.vstack(X_train_list)
y_train = np.concatenate(y_train_list)
groups_train = np.array(groups_train)
group_labels_train = np.array(group_labels_train)

# Generate validation data
np.random.seed(1234)
n_queries_val = int(n_queries_train*0.2) # number of events in validation table

X_val_list = []
y_val_list = []
groups_val = []
group_labels_val = []

for i in range(n_queries_val):
    group = 0 if i < int(n_queries_val*0.8) else 1  # 80-20 split
    group_labels_val.append(group)
    
    n_docs = np.random.randint(5, 15)
    groups_val.append(n_docs)
    
    X_query = np.random.randn(n_docs, n_features)
    y_query = np.random.randint(0, 5, n_docs)
    
    X_val_list.append(X_query)
    y_val_list.append(y_query)

X_val = np.vstack(X_val_list)
y_val = np.concatenate(y_val_list)
groups_val = np.array(groups_val)
group_labels_val = np.array(group_labels_val)

np.random.seed(424)  # Reset seed

print(f"\nTraining Data: {len(groups_train)} queries, {len(X_train)} instances")
print(f"Validation Data: {len(groups_val)} queries, {len(X_val)} instances")

tmpdir = "" # provide directory to store the files

# Write training data files
train_data_file = os.path.join(tmpdir, "train.txt")
train_group_file = os.path.join(tmpdir, "train.group")

# Write validation data files
val_data_file = os.path.join(tmpdir, "val.txt")
val_group_file = os.path.join(tmpdir, "val.group")

# Write training libsvm format
with open(train_data_file, 'w') as f:
    idx = 0
    for query_idx, group_size in enumerate(groups_train):
        for _ in range(group_size):
            features = ' '.join([f'{j}:{X_train[idx, j]:.6f}' for j in range(n_features)])
            f.write(f'{y_train[idx]} {features}\n')
            idx += 1

# Write validation libsvm format
with open(val_data_file, 'w') as f:
    idx = 0
    for query_idx, group_size in enumerate(groups_val):
        for _ in range(group_size):
            features = ' '.join([f'{j}:{X_val[idx, j]:.6f}' for j in range(n_features)])
            f.write(f'{y_val[idx]} {features}\n')
            idx += 1

# Write training group file
with open(train_group_file, 'w') as f:
    for g in groups_train:
        f.write(f'{g}\n')

# Write validation group file
with open(val_group_file, 'w') as f:
    for g in groups_val:
        f.write(f'{g}\n')

print(f"Files written to: {tmpdir}")
print(f"  Training: {train_data_file}, {train_group_file}")
print(f"  Validation: {val_data_file}, {val_group_file}")

### Model 1: Without any weights
# Load DMatrix from files
dtrain_file = xgb.DMatrix(f'{train_data_file}?format=libsvm')
dval_file = xgb.DMatrix(f'{val_data_file}?format=libsvm')

# Set groups from files
train_groups_loaded = np.loadtxt(train_group_file, dtype=int)
val_groups_loaded = np.loadtxt(val_group_file, dtype=int)

dtrain_file.set_group(train_groups_loaded)
dval_file.set_group(val_groups_loaded)

print(f"\nLoaded training: {len(dtrain_file.get_group())}")
print(f"  Weight: {dtrain_file.get_weight()}") # this will be empty as expected

print(f"\nLoaded Validation: {len(dval_file.get_group())}")
print(f"  Weight: {dval_file.get_weight()}") # this will be empty as expected

# Train
params = {
    'objective': 'rank:ndcg',
    'eval_metric': 'ndcg@5',
    'tree_method': 'hist',
    'eta': 0.2,
    'max_depth': 6,
    'seed': 42
}

model_baseline = xgb.train(params, dtrain_file, num_boost_round=100, verbose_eval=False)

# inferring on the validation dataset
preds_baseline = model_baseline.predict(dval_file)
group_ptr_baseline = dval_file.get_uint_info('group_ptr')
labels = dval_file.get_label()

results_baseline = {0: [], 1: []}
for i in range(len(val_groups_loaded)):
    start_idx = group_ptr_baseline[i]
    end_idx = group_ptr_baseline[i + 1]

    query_labels = labels[start_idx:end_idx]
    query_preds = preds_baseline[start_idx:end_idx]

    if np.max(query_labels) > 0:
        try:
            ndcg = ndcg_score([query_labels], [query_preds], k=5)
            results_baseline[group_labels_val[i]].append(ndcg)
        except:
            pass

print(f"\nResults (no weights):")
print(f"  Group 0 (80%): NDCG@5 = {np.mean(results_baseline[0]):.6f}")
print(f"  Group 1 (20%): NDCG@5 = {np.mean(results_baseline[1]):.6f}")

### Model 2: With extreme weights

# Save the weights
train_weight_file = os.path.join(tmpdir, "train.weight")
val_weight_file = os.path.join(tmpdir, "val.weight")

# Write training weight file (extreme weights)
train_weights_extreme = np.where(group_labels_train == 0, 1.0, 100.0)
with open(train_weight_file, 'w') as f:
    for w in train_weights_extreme:
        f.write(f'{w}\n')

# Write validation weight file (extreme weights)
val_weights_extreme = np.where(group_labels_val == 0, 1.0, 100.0)
with open(val_weight_file, 'w') as f:
    for w in val_weights_extreme:
        f.write(f'{w}\n')

# Load DMatrix from files
dtrain_file = xgb.DMatrix(f'{train_data_file}?format=libsvm')
dval_file = xgb.DMatrix(f'{val_data_file}?format=libsvm')

# Set groups and weights from files
train_groups_loaded = np.loadtxt(train_group_file, dtype=int)
train_weights_loaded = np.loadtxt(train_weight_file, dtype=float)
val_groups_loaded = np.loadtxt(val_group_file, dtype=int)
val_weights_loaded = np.loadtxt(val_weight_file, dtype=float)

print(f"\nLoaded training: {len(train_groups_loaded)} groups, {len(train_weights_loaded)} weights")
print(f"  Weight range: [{np.min(train_weights_loaded)}, {np.max(train_weights_loaded)}]")

print(f"Loaded validation: {len(val_groups_loaded)} groups, {len(val_weights_loaded)} weights")
print(f"  Weight range: [{np.min(val_weights_loaded)}, {np.max(val_weights_loaded)}]")

dtrain_file.set_group(train_groups_loaded)
dtrain_file.set_weight(train_weights_loaded)

dval_file.set_group(val_groups_loaded)
dval_file.set_weight(val_weights_loaded)

# Train
params = {
    'objective': 'rank:ndcg',
    'eval_metric': 'ndcg@5',
    'tree_method': 'hist',
    'eta': 0.2,
    'max_depth': 6,
    'seed': 42
}

model_file = xgb.train(params, dtrain_file, num_boost_round=100, verbose_eval=True)

# Evaluate on the validation data set by group
preds = model_file.predict(dval_file)
labels = dval_file.get_label()
group_ptr = dval_file.get_uint_info('group_ptr')

# validate that the validation dataset has weights
print(dval_file.get_weight())

results_file = {0: [], 1: []}
for i in range(len(val_groups_loaded)):
    start_idx = group_ptr[i]
    end_idx = group_ptr[i + 1]

    query_labels = labels[start_idx:end_idx]
    query_preds = preds[start_idx:end_idx]

    if np.max(query_labels) > 0:
        try:
            ndcg = ndcg_score([query_labels], [query_preds], k=5)
            results_file[group_labels_val[i]].append(ndcg)
        except:
            pass

print(f"\nResults (with weights):")
print(f"  Group 0 (80%, weight=1.0): NDCG@5 = {np.mean(results_file[0]):.6f}")
print(f"  Group 1 (20%, weight=100.0): NDCG@5 = {np.mean(results_file[1]):.6f}")

```
Output:

```
Results (no weights):
  Group 0 (80%): NDCG@5 = 0.635597
  Group 1 (20%): NDCG@5 = 0.647720

Results (with weights):
  Group 0 (80%, weight=1.0): NDCG@5 = 0.635265
  Group 1 (20%, weight=100.0): NDCG@5 = 0.651240
```

Even with extreme weights for group 1, the nDCG with weights is very similar to nDCG without weights. 

Can you please help me understand if there is something that I am missing? Thanks

**Comment by HemakshiShardha** [comment]
@trivialfis just checking to see if you got a chance to look at the above example? Let me know if you need any clarification. Thanks

**Comment by trivialfis** [comment]
Apologies for the slow reply. I will look into it next week.

**Comment by HemakshiShardha** [comment]
@trivialfis Following up on the above

**Comment by trivialfis** [comment]
Got a bit confused by what the script is trying to do, will take another look.

**Comment by HemakshiShardha** [comment]
Here is a quick run-down of what the script is trying to do:•  

Creates synthetic ranking data for two “departments”:
- 80% of queries belong to Group 0 (majority group)
- 20% belong to Group 1 (minority group)

Simulates ranking datasets (training + validation) in LibSVM format with group information for ranking (.group files).
- Each query has 5–15 “documents” with random features and random relevance scores (0–4).

Trains two XGBoost ranking models (rank:ndcg):
- Model 1 (Baseline): No weights.
- Model 2 (Weighted): Each query in the minority group gets a 100× higher weight.

Evaluates both models separately for each group using sklearn.metrics.ndcg_score and finds that even with extreme weights, the nDCG has barely changed.

**Comment by trivialfis** [comment]
The effect of weight is not clear-cut. For LTR models, it's used in two places:
- Affect the quantization (bucketing the data). Each histogram bin strives to have similar *sum* weights to other bins. An instance with a larger weight occupies a larger portion of a histogram bin, sort of. The histogram is used to approximate the gain function by discretizing the feature values.
- Affect the gradient calculation. The group weight is used to scale the instance gradient.

Rest is the same as without weight. Finding a tree split based on the sum gradient, etc. The effect of weight is less clear than, say, logistic regression.

I ran your script and can confirm that the weight is used (I looked inside XGBoost). As for why the end NDCG result doesn't change much, I cannot answer without stepping through how the tree is split.

For reference, the gradient scaling happens here: https://github.com/dmlc/xgboost/blob/bd54840ad06f89195435f684a3101c8d3f72f733/src/objective/lambdarank_obj.cc#L249

**Comment by HemakshiShardha** [comment]
Noted. Thanks for looking into this and confirming that weights are being used.
